ide/coqide/coqOps.ml

```diff
@@ -494,11 +494,17 @@ object(self)
 
   method private attach_tooltip ?loc sentence text =
     let start_sentence, stop_sentence, phrase = self#get_sentence sentence in
-    let pre_bytes, post_bytes = Option.cata Loc.unloc (0, String.length phrase) loc in
-    let pre = b2c phrase (pre_bytes - (sentence.byte_start)) in
-    let post = b2c phrase (post_bytes - (sentence.byte_start)) in
-    let start = start_sentence#forward_chars pre in
-    let stop = start_sentence#forward_chars post in
+    let start, stop = match loc with
+      | None -> start_sentence, stop_sentence
+      | Some loc ->
+        (* FIXME: tooltips don't show correctly for this case *)
```

Comment:
I fail to see the point of this PR over #16978 , for the `None` case the code is the same, for `Some` case the code is wrong, whereas the on in #16978 should be correct TTBOMK.

Reply:
@ejgallego The explanation provided by Jim is that your last commit in your PR reintroduces a performance issue:

>The last commit ("Fix bizarre location-handling code.") reintroduces the O(n^2) behavior. The ones before it are OK. (Pff.v, a 1.2 MB file, takes about 3 minutes to run in CoqIDE both on master and with the first 3 commits. With the 4th commit, I gave up after 10 minutes, and it had only processed half the source.)

So the point of this PR is supposed to be: fix the tooltip issue without reintroducing the quadratic behavior.

Note that I'm just commenting based on my reading of the comments, I haven't looked at the actual code on either PR.

Reply:
I'll have a look into the performance issue, but it is not introduced in the code part this PR is touching, the last commit in the PR does indeed touch 2 different code paths.

I'll split the commit in two. This PR still makes little sense as we have better code for the `Some` path in the other PR, whereas this one just puts some code that looks pretty strange.

Reply:
> I fail to see the point of this PR

The issue is unrelated to your original PR.  I took the part of your changes that made sense.  Currently, no warning uses the code for the `Some` branch, and IIRC your code was not correct for the `Some` case.  I assumed it was urgent to get your other code into 8.17 and that we should fix the tooltip problem.  Fixing the `Some` case doesn't have to be in 8.17, though I planned to look at it very soon.  Now I'm not sure I should bother.

Reply:
My code was almost correct for the `Some` case, it only did mishandle the case where UTF-8 symbols where before in the buffer, whereas the current code here mishandles all of them. Moreover it was noted as a fixme which is now fixed.

Indeed, why you would like to spend time fixing something that is already fixed is an interesting question.

By the way, you should be able to hit the `Some` case, at least quite a few warnings do set `?loc` , then why the feedback may be discarded at the STM level is a different issue.